// @mrwconsulting/shinto/lib/shinto/codebuild-helper.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBuildHelper = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const pipelines = require("aws-cdk-lib/pipelines");
const constants_1 = require("./constants");
const types_1 = require("./types");
/**
 *
 * @class
 * @description - CodeBuildStep Helper
 *
 */
class CodeBuildHelper {
    constructor(props) {
        this._props = {
            ...props,
        };
    }
    /**
     *
     * @description - java build pipeline
     * @param props CodeBuildStep properties
     * @returns CodeBuildStep
     *
     */
    addJavaBuildStep(props) {
        let id = types_1.BuildType.JAVA;
        let codeBuildStepProps = {
            ...props,
            commands: [
                ...this._props.buildSpec.commands || constants_1.ShintoConstants.JAVA_BUILD_COMMANDS,
            ],
            installCommands: [
                `cd ${this._props.metadata?.PIPELINE_WORKSPACE || constants_1.ShintoConstants.DEFAULT_DIR}`,
                ...this._props.buildSpec.installCommands || constants_1.ShintoConstants.JAVA_BUILD_INSTALL_COMMANDS,
            ],
            env: {
                ...props.env,
                ...this._props.buildSpec.env,
                PACKAGE_MANAGER: `${this._props.buildSpec.packageManager || types_1.PackageManagerType.MAVEN}`,
            },
        };
        let codeBuildStep = new pipelines.CodeBuildStep(id, codeBuildStepProps);
        return codeBuildStep;
    }
    /**
     *
     * @description - python build pipeline
     * @param props CodeBuildStep properties
     * @returns CodeBuildStep
     *
     */
    addPythonBuildStep(props) {
        let id = types_1.BuildType.PYTHON;
        let codeBuildStepProps = {
            ...props,
            commands: [
                ...this._props.buildSpec.commands || constants_1.ShintoConstants.PYTHON_BUILD_COMMANDS,
            ],
            installCommands: [
                `cd ${this._props.metadata?.PIPELINE_WORKSPACE || constants_1.ShintoConstants.DEFAULT_DIR}`,
                ...this._props.buildSpec.installCommands || constants_1.ShintoConstants.PYTHON_BUILD_INSTALL_COMMANDS,
            ],
            env: {
                ...props.env,
                ...this._props.buildSpec.env,
                PYTEST_DIR: `${this._props.metadata?.PYTEST_DIR || constants_1.ShintoConstants.PYTHON_PYTEST_DIR}`,
                PACKAGE_MANAGER: `${this._props.buildSpec.packageManager || types_1.PackageManagerType.PIP}`,
            },
        };
        let codeBuildStep = new pipelines.CodeBuildStep(id, codeBuildStepProps);
        return codeBuildStep;
    }
}
exports.CodeBuildHelper = CodeBuildHelper;
_a = JSII_RTTI_SYMBOL_1;
CodeBuildHelper[_a] = { fqn: "@mrwconsulting/shinto.CodeBuildHelper", version: "0.0.2" };
