// @mrwconsulting/shinto/lib/shinto/types.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CodeBuildStepType = exports.BuildType = exports.PackageManagerType = exports.CommandType = exports.RegionType = exports.EnvironmentType = void 0;
/**
 *
 * @description - Supported Environment Type(s)
 *
 */
var EnvironmentType;
(function (EnvironmentType) {
    EnvironmentType["DEV"] = "DEV";
    EnvironmentType["SIT"] = "SIT";
    EnvironmentType["PVE"] = "PVE";
    EnvironmentType["PROD"] = "PROD";
})(EnvironmentType = exports.EnvironmentType || (exports.EnvironmentType = {}));
/**
 *
 * @description - Supported AWS Regions
 *
 */
var RegionType;
(function (RegionType) {
    RegionType["US_EAST_1"] = "us-east-1";
    RegionType["US_WEST_2"] = "us-west-2";
})(RegionType = exports.RegionType || (exports.RegionType = {}));
/**
 *
 * @description - Supported Command Type(s)
 *
 */
var CommandType;
(function (CommandType) {
    CommandType["COMMANDS"] = "COMMANDS";
    CommandType["INSTALL_COMMANDS"] = "INSTALL_COMMANDS";
})(CommandType = exports.CommandType || (exports.CommandType = {}));
/**
 *
 * @description - Supported Package Management Type(s)
 *
 */
var PackageManagerType;
(function (PackageManagerType) {
    PackageManagerType["GRADLE"] = "GRADLE";
    PackageManagerType["MAVEN"] = "MAVEN";
    PackageManagerType["PIP"] = "PIP";
    PackageManagerType["POETRY"] = "POETRY";
})(PackageManagerType = exports.PackageManagerType || (exports.PackageManagerType = {}));
/**
 *
 * @description - Supported Build Type(s)
 *
 */
var BuildType;
(function (BuildType) {
    BuildType["JAVA"] = "JAVA";
    BuildType["PYTHON"] = "PYTHON";
})(BuildType = exports.BuildType || (exports.BuildType = {}));
/**
   *
   * @description - Combined Enum Type(s)
   *
   */
exports.CodeBuildStepType = {
    ...BuildType,
};
