// @mrwconsulting/shinto/lib/api/index.js
"use strict";
var _a;
Object.defineProperty(exports, "__esModule", { value: true });
exports.ShintoManagerApi = void 0;
const JSII_RTTI_SYMBOL_1 = Symbol.for("jsii.rtti");
const path = require("path");
const aws_cdk_lib_1 = require("aws-cdk-lib");
const aws_apigateway_1 = require("aws-cdk-lib/aws-apigateway");
const aws_lambda_1 = require("aws-cdk-lib/aws-lambda");
const aws_ssm_1 = require("aws-cdk-lib/aws-ssm");
const constants_1 = require("./constants");
class ShintoManagerApi {
    constructor(props) {
        this._app = new aws_cdk_lib_1.App();
        this._stack = new aws_cdk_lib_1.Stack(this._app, 'shinto-manager-stack'.toLowerCase(), {
            env: {
                region: props.region || constants_1.ShintoManagerConstants.DEFAULT_REGION,
                account: props.account,
            },
        });
        this._api = new aws_apigateway_1.RestApi(this._stack, 'shinto-manager-api');
    }
    stack() {
        const register_function = new aws_lambda_1.DockerImageFunction(this._stack, 'register_function', {
            code: aws_lambda_1.DockerImageCode.fromImageAsset(path.join(__dirname, 'function', 'register')),
        });
        const unregister_function = new aws_lambda_1.DockerImageFunction(this._stack, 'unregister_function', {
            code: aws_lambda_1.DockerImageCode.fromImageAsset(path.join(__dirname, 'function', 'unregister')),
        });
        const register = this._api.root.addResource('register');
        register.addMethod('POST');
        register.addMethod('POST', new aws_apigateway_1.LambdaIntegration(register_function, {}));
        const unregister = this._api.root.addResource('unregister');
        unregister.addMethod('POST');
        unregister.addMethod('POST', new aws_apigateway_1.LambdaIntegration(unregister_function, {}));
        let api_url = new aws_ssm_1.StringParameter(this._stack, 'api_url', {
            parameterName: constants_1.ShintoManagerConstants.API_URL,
            stringValue: this._api.url,
        });
        console.log('API Url: ' + api_url.stringValue);
        return this._stack;
    }
}
exports.ShintoManagerApi = ShintoManagerApi;
_a = JSII_RTTI_SYMBOL_1;
ShintoManagerApi[_a] = { fqn: "@mrwconsulting/shinto.ShintoManagerApi", version: "0.0.2" };
